test-network/scripts/add_new_orderer_to_config.py

```diff
@@ -0,0 +1,98 @@
+import argparse
+import base64
+import copy
+import json
+import math
+from typing import Any
+
+
+def parse_args():
+    parser = argparse.ArgumentParser(
+        prog='Config Update',
+        description='What the program does',
+        epilog='Text at the bottom of help')
+    parser.add_argument('config_path', type=str)
+    parser.add_argument('updated_config_path', type=str)
+    parser.add_argument('-a', '--address', type=str, required=True)
+    parser.add_argument('-i', '--identity', type=str, required=True)
+    parser.add_argument('-s', '--server-cert', type=str, required=True)
+    parser.add_argument('-c', '--client-cert', type=str, required=True)
+    return parser.parse_args()
+
+
+def _pem_file_to_base64(path: str) -> str:
+    with open(path, 'rb') as binary_file:
+        binary_file_data = binary_file.read()
+        base64_encoded_data = base64.b64encode(binary_file_data)
+        return base64_encoded_data.decode('utf-8')
+
+
+def _log_update(name: str, old: Any, new: Any) -> None:
+    print('=' * 50)
+    print(f'Updating {name}:')
+    print(f'{old}')
+    print(">" * 25)
+    print(f'{new}')
+    print('=' * 50)
+
+
+def _calculate_bft_quorum(n: int) -> int:
+    f = int((n - 1) / 3)
+    return int(math.ceil((n + f + 1) / 2))
+
+
+def update_config(config_path: str, updated_config_path: str, address: str, identity_pem_path: str, server_pem_path: str, client_pem_path: str):
+    with open(config_path, 'r') as f:
+        config = json.load(f)
+    identity = _pem_file_to_base64(identity_pem_path)
+    client_cert = _pem_file_to_base64(client_pem_path)
+    server_cert = _pem_file_to_base64(server_pem_path)
+    host, port = address.split(':')
+
+    addresses = config['channel_group']['groups']['Orderer']['groups']['OrdererOrg']['values']['Endpoints']['value']['addresses']
+    addresses_before_update = copy.deepcopy(addresses)
+    original_orderers_count = len(addresses_before_update)
+    addresses.append(f'{addresses[0].split(":")[0]}:{port}')
+    new_orderers_count = len(addresses)
+    _log_update('addresses', addresses_before_update, addresses)
+
+    identities = config['channel_group']['groups']['Orderer']['policies']['BlockValidation']['policy']['value']['identities']
+    identities_before_update = copy.deepcopy(identities)
+    new_identity = copy.deepcopy(identities[0])
+    new_identity['principal']['id_bytes'] = identity
+    identities.append(new_identity)
+    _log_update('block validation identities', identities_before_update, identities)
+
+    rule = config['channel_group']['groups']['Orderer']['policies']['BlockValidation']['policy']['value'][
+        'rule']
+    rule_before_update = copy.deepcopy(rule)
+    rule['n_out_of']['n'] = _calculate_bft_quorum(new_orderers_count)
+    rule['n_out_of']['rules'].append({'signed_by': new_orderers_count - 1})
+    _log_update('block validation rules', rule_before_update, rule)
```

Comment:
One general question and one specific question.

General question - I didn't see this file originally... is it used in this PR? Or will other instructions use it?

Specific question - I didn't see anything like these BlockValidation policies in the original configtx.yaml, like I did for the Orderer Endpoints (above) and consenter mapping (below). Can you educate me on this... why are they not in configtx.yaml? Is it in the channel config implicitly somehow originally? Is this new for BFT?

Reply:
> I didn't see anything like these BlockValidation policies in the original configtx.yaml, like I did for the Orderer Endpoints (above) and consenter mapping (below). Can you educate me on this... why are they not in configtx.yaml? Is it in the channel config implicitly somehow originally? Is this new for BFT?


I'll answer this one, and will let Arkadi answer the first one.

First of all, remember that the "name" of the policy in configtx.yaml is a mere pointer to a meta policy of "any orderer". 
The meta policies we have, or, the policies we have in general in Fabric, are not suitable to define BFT policies, as BFT policies depend on additional input found in the channel config. To add support for that, it requires invasive changes to the channel config infrastructure. 

However, a signature policy is a supported policy that can always be used and it is flexible enough to fit BFT block validation.

In the spirit of making Fabric more generic and pluggable, we don't want to define policy names for every type of consensus we can imagine. Some protocols require more signatures, some require less, some require a single threshold signature, etc. 
While in Raft it's easy to define a policy string ("any orderer"), for BFT it's complex to do that in the configtx.yaml, and it can also be error prone. 

For BFT, the Fabric tooling (configtxgen) overrides the block validation policy that is defined in the configtx.yaml, both in order to "protect the user from a misconfiguration", and also because it was my belief that a user should not need to understand the underlying BFT protocol to configure the policy. 
So, when a channel is created using configtxgen, the genesis block contains the right policy. When a channel is updated, the orderer itself checks that the policy was correctly computed via an [ad-hoc verification](https://github.com/hyperledger/fabric/blob/main/orderer/consensus/smartbft/configverifier.go#L93-L168).

Reply:
For the general question:
This file is used in the following tutorial:
https://github.com/hyperledger/fabric/pull/4390